### PR TITLE
Add Java service implementation for eth_call request

### LINKED ISSUE #4463: Implement REST API for `eth_call` JSON-RPC request
### Problem

Mirror nodes will start support of `eth_call` JSON-RPC calls coming from the JSON-RPC Relay. In order to fulfil these needs we should build the needed infrastructure for a REST API.

### Solution

Implement the following subtasks:

- [x] #4465 
- [x] #4466 
- [x] #4467 
- [x] #4468
- [x] #4503 
- [x] #4873 
- [x] #4894 

### Alternatives

_No response_

### LINKED ISSUE #4467: Add Java service for `eth_call` and `eth_estimateGas` requests
### Problem

We need the support of `eth_call` and `eth_estimateGas` requests in `JSON-RPC relay`. To do so, a Java based service should be implemented that follows mirror-node standard for request and return bodies and fulfil the needs of `eth_call` and `eth_estimateGas` requests. To differentiate between the two types of calls, a query parameter flag called `estimate` would be used.

### Solution

Define a Java service under `/api/v1/contracts/call` mapping. It would use the `hedera-evm-api` library to perform a transaction. The returned result from the library would be encapsulated in `HederaEvmTransactionProcessingResult`. It would be used to fill the return body of the service that will be ingested by the relay. A query parameter called `estimate` would help us know which of two defined types (`result` and `gas`) to be returned.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by nikolovyanko**
Introducing a [ContractCallService](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/eth/ContractCallService.java) service class used by the [ContractController](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java).

[ContractCallService](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/eth/ContractCallService.java) depend on a [MirrorEvmTxProcessorFacade](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessorFacadeImpl.java) that constructs needed by the [MirrorEvmTxProcessor](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessor.java) dependancies and triggers transactional [execute](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessor.java#L69).

**Related issue(s)**:

Fixes #4467 

**Notes for reviewer**:

> [PricesAndFeesImpl](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java) is a hardcoded implementation which will be extended with the introduction of estimate_gas functionality.

> [GasCalculator](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessorFacadeImpl.java#L54) will also be addressed with the estimate_gas PR

> [mcps,ccps](https://github.com/hashgraph/hedera-mirror-node/blob/fe1e2c8411841293f849c0a61d95c59bdcfc9ee3/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessorFacadeImpl.java#L56)` are populated manually rather than injected since the evm-module-lib lack such logic and this needs to be discussed further with besu.

### CODE REVIEW

**Comment by steven-sheehy** [comment]
Can we close https://github.com/hashgraph/hedera-mirror-node/pull/4633 and delete its branch in favor of this PR?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessorFacadeImpl.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import com.hedera.node.app.service.evm.store.contracts.AbstractCodeCache;
+import com.hedera.node.app.service.evm.store.contracts.HederaEvmMutableWorldState;
+import com.hedera.node.app.service.evm.store.contracts.HederaEvmWorldState;
+
+
+@Named
+public class MirrorEvmTxProcessorFacadeImpl implements MirrorEvmTxProcessorFacade {
+    private final MirrorEvmTxProcessor processor;
+
+    public MirrorEvmTxProcessorFacadeImpl(
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/MirrorEvmTxProcessorFacadeImpl.java:52]: Probably we can make all arguments `final`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java [hunk 2]
```diff
@@ -20,9 +20,20 @@
...
-public class MirrorOperationTracer implements HederaEvmOperationTracer {
+import com.hedera.node.app.service.evm.contracts.execution.PricesAndFeesProvider;
 
+@Named
+@NoArgsConstructor
+public class PricesAndFeesImpl implements PricesAndFeesProvider {
 
+    @Override
+    public long currentGasPrice(Instant now, HederaFunctionality function) {
+        return 1000L;
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java:36]: We can add comment like:
`//FEATURE WORK - precise gas price calculation to be provided with eth_estimateGas implementation`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java [hunk 3]
```diff
@@ -20,9 +20,20 @@
...
 
-public class MirrorOperationTracer implements HederaEvmOperationTracer {
+import com.hedera.node.app.service.evm.contracts.execution.PricesAndFeesProvider;
 
+@Named
+@NoArgsConstructor
+public class PricesAndFeesImpl implements PricesAndFeesProvider {
 
+    @Override
+    public long currentGasPrice(Instant now, HederaFunctionality function) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java:35]: We can make arguments `final`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        final var processingResult =
+                mirrorEvmTxProcessor.execute(
+                        body.getSender(),
+                        body.getReceiver(),
+                        body.getProvidedGasLimit(),
+                        body.getValue(),
+                        body.getCallData(),
+                        body.isStatic());
+
+        final var status = getStatusOrDefault(processingResult);
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:56]: I think we can get rid of `ResponseCodeUtil` to reduce `mirror-node` complexity and we can directly check whether the `result.isSuccessful` - if not - we throw an `InvalidTransactionException`, without the need of matching specific error response code.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java [hunk 5]
```diff
@@ -35,23 +41,66 @@
...
+    private final ContractCallService contractCallService;
 
     @PostMapping(value = "/call")
     Mono<ContractCallResponse> call(@RequestBody @Valid ContractCallRequest request) {
-        throw new UnsupportedOperationException(NOT_IMPLEMENTED_ERROR);
+        if (request.isEstimate()) {
+            throw new UnsupportedOperationException(NOT_IMPLEMENTED_ERROR);
+        }
+        //make sure there is a valid sender for value transfers
+        if (request.getValue() > 0 && request.getFrom() == null) {
```
Comment (Nana-EC) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java:67]: Can we move these lines into `ContractCallRequest.isValidValueTransfer()`
```suggestion
        if (!request.isValidValueTransfer())
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+@Named
+@RequiredArgsConstructor
+public class ContractCallService {
+    private final MirrorEvmTxProcessorFacade mirrorEvmTxProcessor;
+
+    public String processCall(CallServiceParameters body) {
+        final var processResult = doProcessCall(body);
+
+        return processResult.getOutput().toHexString();
```
Comment (Nana-EC) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:43]: Q: is there a chance for `processResult.getOutput()` to be null?
Is output always in hex form?
If so move that casting into the `getOutput()` setting.

**Review by Nana-EC — COMMENTED** [review]
Some suggestions

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallRequest.java [hunk 7]
```diff
@@ -59,4 +59,8 @@ public class ContractCallRequest {
 
     @Min(0)
     private long value;
+
+    public boolean isInvalidValueTransfer(){
+        return (value > 0 && from == null);
+    }
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallRequest.java:65]: Use validation [groups](https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#_requesting_groups) for conditional validations. This avoid having to write conditional validation logic and continue to depend upon annotations. It also provides better error messages to clients since it does all validation at once and returns all errors instead of requiring separate requests to discover the conditional logic.

```java
    @Hex(minLength = ADDRESS_LENGTH, maxLength = ADDRESS_LENGTH)
    @NotNull(groups = TransferCheck.class)
    private String from;

    @Min(0)
    @Min(value = 1, groups = TransferCheck.class)
    private long value;
```
```java
public interface TransferCheck {
}
```

Have to switch controller from `@Valid` to the below to make it work:
```java
@Validated({Default.class, TransferCheck.class})
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallResponse.java [hunk 8]
```diff
@@ -20,9 +20,8 @@
...
 
-import lombok.Value;
-
-@Value
-public class ContractCallResponse {
-    String result;
+public record ContractCallResponse(String result) {
+    public static ContractCallResponse of(String response) {
+        return new ContractCallResponse(response);
+    }
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallResponse.java:26]: I don't think this method provides any value over just calling the record-generated constructor.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 9]
```diff
@@ -50,6 +48,8 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @Max(100)
     private int maxGasRefundPercentage = 20;
 
+    private int expirationCacheTime = 600;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:51]: It might be a change upstream, but can we use `Duration` to model time-based properties`? It allows our config to accept variable units like `5m` or `600s`. It also makes it clear what unit this integer is in. Internally you can convert to integer to send to the library.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.hederahashgraph.api.proto.java.HederaFunctionality;
+import java.time.Instant;
+import javax.inject.Named;
+import lombok.NoArgsConstructor;
+
+import com.hedera.node.app.service.evm.contracts.execution.PricesAndFeesProvider;
+
+@Named
+@NoArgsConstructor
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/contracts/execution/PricesAndFeesImpl.java:31]: nit: Can remove since Java provides a no-arg constructor by default.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/account/MirrorEvmContractAliases.java [hunk 11]
```diff
@@ -20,24 +20,25 @@
...
 
-    public MirrorEvmContractAliases(MirrorEntityAccess mirrorEntityAccess) {
-        this.mirrorEntityAccess = mirrorEntityAccess;
-    }
 
     @Override
     public Address resolveForEvm(Address addressOrAlias) {
         final var entity = mirrorEntityAccess.findEntity(addressOrAlias);
-        return Address.wrap(Bytes.wrap(entity.map(Entity::getEvmAddress).orElse(addressOrAlias.toArray())));
+        return Address.wrap(Bytes.wrap(entity.map(Entity::getEvmAddress).orElse(new byte[0])));
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/account/MirrorEvmContractAliases.java:42]: Should use a constant with a readable name.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                mirrorEvmTxProcessor.execute(
+                        body.getSender(),
+                        body.getReceiver(),
+                        body.getProvidedGasLimit(),
+                        body.getValue(),
+                        body.getCallData(),
+                        body.isStatic());
+
+        if (!processingResult.isSuccessful()) {
+            throw new InvalidTransactionException(INVALID_TRANSACTION);
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:58]: Can we not provide more detail as to why the transaction is invalid?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        final var txnResult = doProcessCall(body);
+        final var callResult = txnResult.getOutput() != null
+                ? txnResult.getOutput() : Bytes.EMPTY;
+
+        return callResult.toHexString();
+    }
+
+    private HederaEvmTransactionProcessingResult doProcessCall(final CallServiceParameters body) {
+        final var processingResult =
+                mirrorEvmTxProcessor.execute(
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:49]: Does this processor catch all exceptions and return a processingResult regardless or should this be wrapped in a try/catch?

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/controller/ContractControllerTest.java [hunk 14]
```diff
@@ -181,9 +190,24 @@ void callValidBlockType(String value) {
                 .body(BodyInserters.fromValue(request))
                 .exchange()
                 .expectStatus()
-                .isEqualTo(NOT_IMPLEMENTED)
-                .expectBody(GenericErrorResponse.class)
-                .isEqualTo(new GenericErrorResponse(NOT_IMPLEMENTED_ERROR));
+                .isEqualTo(OK);
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/controller/ContractControllerTest.java:193]: estimateGas=true should return NOT_IMPLEMENTED, right?

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/controller/ContractControllerTest.java [hunk 15]
```diff
@@ -181,9 +190,24 @@ void callValidBlockType(String value) {
...
+    }
+
+    @Test
+    void callSuccess(){
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/controller/ContractControllerTest.java:197]: Need a test for invalid value transfer parameters like value > 0 and value = null.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+            new HederaBlockValues(120000000, 1030,
+            Instant.now());
+    @Resource
+    private ContractCallService contractCallService;
+
+    @MockBean
+    private MirrorEntityAccess mirrorEntityAccess;
+
+    @MockBean
+    private StaticBlockMetaSource blockMetaSource;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java:59]: We need at least one integration test that invokes the actual EVM and doesn't mock anything out.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    void processCallSuccess() {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java:62]: Needs negative tests.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java [hunk 18]
```diff
@@ -61,9 +110,24 @@ private Mono<GenericErrorResponse> unsupportedOpResponse(UnsupportedOperationExc
     @ExceptionHandler
     @ResponseStatus(BAD_REQUEST)
     private Mono<GenericErrorResponse> validationError(WebExchangeBindException e) {
+        log.warn("Validation error: {}", e.getMessage());
         return errorResponse(parseValidationError(e));
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java]: Not completed. Need separate error items in the list.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/exception/ResponseCodeUtil.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+            final HederaEvmTransactionProcessingResult result) {
+        if (result.isSuccessful()) {
+            return SUCCESS;
+        }
+        var maybeHaltReason = result.getHaltReason();
+        if (maybeHaltReason.isPresent()) {
+            var haltReason = maybeHaltReason.get();
+            if (HederaExceptionalHaltReason.SELF_DESTRUCT_TO_SELF == haltReason) {
+                return ResponseCodeEnum.OBTAINER_SAME_CONTRACT_ID;
+            } else if (HederaExceptionalHaltReason.INVALID_SOLIDITY_ADDRESS == haltReason) {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/exception/ResponseCodeUtil.java:78]: As codecov suggests, we're missing code coverage for these branches. Preferably in the service layer and not just util tests. But I'll take what I can get.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+            return
+                    mirrorEvmTxProcessor.execute(
+                            body.getSender(),
+                            body.getReceiver(),
+                            body.getProvidedGasLimit(),
+                            body.getValue(),
+                            body.getCallData(),
+                            body.isStatic());
+
+        } catch (Exception e) {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:64]: This swallows the exception. Shouldn't this return some response to the user or at minimum log the error?

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+class ContractCallServiceTest extends Web3IntegrationTest {
+
+    private static final Bytes RUNTIME_BYTES_VIEW_FUNC = Bytes.fromHexString(
+            "608060405234801561001057600080fd5b50600436106100415760003560e01c80632e64cec1146100465780636057361d146100645780636d4ce63c14610080575b600080fd5b61004e61009e565b60405161005b919061010b565b60405180910390f35b61007e600480360381019061007991906100cf565b6100a7565b005b6100886100b1565b604051610095919061010b565b60405180910390f35b60006033905090565b8060008190555050565b60006048905090565b6000813590506100c981610135565b92915050565b6000602082840312156100e5576100e4610130565b5b60006100f3848285016100ba565b91505092915050565b61010581610126565b82525050565b600060208201905061012060008301846100fc565b92915050565b6000819050919050565b600080fd5b61013e81610126565b811461014957600080fd5b5056fea264697066735822122055a1bf21b71aa5fa69a6d162c173c2e547f0c9263c69887e08891f1d7257ef7464736f6c63430008070033");
+
+    private static final Address CONTRACT_ADDRESS = Address.fromHexString("0x00000000000000000000000000000000000004e2");
+
+    @Resource
+    private ContractCallService contractCallService;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java:48]: Take advantage of the existing `@RequiredArgsConstructor`:

```suggestion
    private final ContractCallService contractCallService;
```

File: buildSrc/src/main/resources/suppressions.xml [hunk 22]
```diff
@@ -28,4 +28,12 @@
         <notes>No snakeyaml release available with a fix yet</notes>
         <cve>CVE-2022-1471</cve>
     </suppress>
+    <suppress>
+        <cve>CVE-2021-37533</cve>
```
Comment (steven-sheehy) [inline buildSrc/src/main/resources/suppressions.xml:32]: This is too broad of a suppression. If this is a false positive you should add the package and a note explaining why it's a false positive. If it's not a false positive and we are relying on a vulnerable Apache Commons Net library then we should not suppress it.

File: buildSrc/src/main/resources/suppressions.xml [hunk 23]
```diff
@@ -28,4 +28,12 @@
...
+    </suppress>
+    <suppress>
+        <notes>hedera-evm-lib</notes>
```
Comment (steven-sheehy) [inline buildSrc/src/main/resources/suppressions.xml:35]: Add a note about why we are suppressing this. Believe it's a false positive to due to confusing another library with the hedera evm library.

File: hedera-mirror-web3/build.gradle.kts [hunk 24]
```diff
@@ -25,9 +25,11 @@ plugins {
 }
 
 dependencies {
+    api("com.hedera.evm:hedera-evm") {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/build.gradle.kts:28]: Why are we using api here? We should use `implementation` since web3 is not a library and shouldn't need to leak this dependency elsewhere.

File: hedera-mirror-web3/pom.xml [hunk 25]
```diff
@@ -125,7 +125,7 @@
         </dependency>
         <dependency>
             <groupId>com.hedera.evm</groupId>
-            <artifactId>hedera-evm-api</artifactId>
```
Comment (steven-sheehy) [inline hedera-mirror-web3/pom.xml:128]: Need to rebase as we've removed Maven.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java [hunk 26]
```diff
@@ -35,23 +40,62 @@
...
+    private CallServiceParameters constructServiceParameters(ContractCallRequest request) {
+        final var fromAddress =
+                request.getFrom() != null
+                        ? Address.fromHexString(request.getFrom())
+                        : Address.ZERO;
+        final var sender = new HederaEvmAccount(fromAddress);
+
+        final var receiver = Address.fromHexString(request.getTo());
+        final var data =
+                request.getData() != null
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java:83]: Need test for empty

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java [hunk 27]
```diff
@@ -61,7 +105,23 @@ private Mono<GenericErrorResponse> unsupportedOpResponse(UnsupportedOperationExc
...
+    @ResponseStatus(BAD_REQUEST)
+    private Mono<GenericErrorResponse> invalidTxnError(InvalidTransactionException e) {
+        log.warn("Transaction error: {}", e.getMessage());
+        return errorResponse(e.getMessage());
+    }
+
+    @ExceptionHandler()
+    @ResponseStatus(INTERNAL_SERVER_ERROR)
+    private Mono<GenericErrorResponse> genericError(Exception e) {
+        log.warn("Generic error: {}", e.getMessage());
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java:123]: We should log at error for internal errors and print the stacktrace.
```suggestion
        log.error("Generic error: ", e);
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/exception/InvalidTransactionException.java [hunk 28]
```diff
@@ -20,12 +20,21 @@
...
+
 public class InvalidTransactionException extends EvmException {
 
     public InvalidTransactionException(final ResponseCodeEnum responseCode) {
         super(responseCode.name());
     }
+
+    public Bytes messageBytes() {
+        final var detail = getMessage();
+        return Bytes.of(detail.getBytes());
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/exception/InvalidTransactionException.java:37]:
```suggestion
        return Bytes.of(detail.getBytes(StandardCharsets.UTF8));
```

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+@Named
+@RequiredArgsConstructor
+public class ContractCallService {
+    private final MirrorEvmTxProcessorFacade mirrorEvmTxProcessor;
+
+    public String processCall(final CallServiceParameters body) {
+        final var txnResult = doProcessCall(body);
+
+        if (!txnResult.isSuccessful()) {
+            throw new InvalidTransactionException(getStatusOrDefault(txnResult));
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:44]: Needs a test case

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+                    mirrorEvmTxProcessor.execute(
+                            body.getSender(),
+                            body.getReceiver(),
+                            body.getProvidedGasLimit(),
+                            body.getValue(),
+                            body.getCallData(),
+                            body.isStatic());
+
+        } catch (Exception e) {
+            return emptyFailProcessingResponse();
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:65]: Needs a test case

**Review by steven-sheehy — CHANGES_REQUESTED** [review]
There's still outstanding from my previous round of feedback. I will be on vacation soon so if we'd like this in and have my sign off changes need to be implemented soon.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallRequest.java [hunk 31]
```diff
@@ -24,13 +24,18 @@
...
 import lombok.Data;
 
 import com.hedera.mirror.web3.convert.BlockTypeDeserializer;
 import com.hedera.mirror.web3.convert.BlockTypeSerializer;
 import com.hedera.mirror.web3.validation.Hex;
 
+import org.hibernate.validator.group.GroupSequenceProvider;
+
 @Data
+@GroupSequenceProvider(TransferValidation.class)
```
[the remark this answers is not in the document]
Comment (nikolovyanko) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/viewmodel/ContractCallRequest.java:38]: I took this approach because with the suggested, otherwise valid case of a call with non-mandatory fields `from=null` and `value=0` fails validation `value field must be greater than or equal to 1`.

**Comment by nikolovyanko** [comment]
> There's still outstanding items from my previous round of feedback. I will be on vacation soon so if we'd like this in and have my sign-off changes need to be implemented soon.

The implementation here was delayed due to a missing logic from the evm-module lib, that will be introduced with this [PR](https://github.com/hashgraph/hedera-services/pull/4477). Some of the  tests in `ContractCallServiceTest` fails as they expect these changes.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java (hunk 20 above)
Comment (nikolovyanko) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:64]: Not sure whether to wrap a potential NullPointer/InvalidArgument.. exceptions returned by the TxnProcessor in `InvalidTransactionException` or to just leave them to be catched by the `ContractController`'s `genericError`?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (nikolovyanko) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java]: Doesn't `parseValidationError(WebExchangeBindException )` fit the requirements?

[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java]: No, it returns a String. It should either return a `Collection<String>`, `List<ErrorMessage>` or the `GenericErrorResponse` with its internal list populated with multiple items.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java (hunk 20 above)
Comment (nikolovyanko) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java:63]: The other 2 possibility  indicating: wrong input, insufficient account balance etc. are `IllegalStateException` and `IllegalArgumentException` everything else I think can be consider as a internal server error and can be be handled by:
https://github.com/hashgraph/hedera-mirror-node/blob/b786289a99fa9d38a65f370f74073035a177cb6e/hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java#L122

Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/service/ContractCallService.java]: Then we should probably catch `IllegalStateException | IllegalArgumentException` and rethrow as `InvalidTransactionException` and not catch others.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java [hunk 32]
```diff
@@ -35,23 +40,62 @@
...
 @RequestMapping("/api/v1/contracts")
+@RequiredArgsConstructor
 @RestController
 class ContractController {
 
     static final String NOT_IMPLEMENTED_ERROR = "Operation not supported yet!";
+    private final ContractCallService contractCallService;
 
     @PostMapping(value = "/call")
     Mono<ContractCallResponse> call(@RequestBody @Valid ContractCallRequest request) {
```
Comment (nikolovyanko) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java:61]: Wondering whether a feature flag for the endpoint is needed? And if yes is an approach with different spring profiles suitable?

Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/controller/ContractController.java:61]: I'm not sure a feature flag is necessary since this module does nothing else besides the contract call at this point, especially after #5099. And we can disable this route via the kubernetes ingress, if needed.

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM! Once we get the latest `hedera-evm` artifact version providing some enhancements, `eth_call` should be working.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
